<h1> Importing Libraries </h1>

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import preprocessing

<h1> Downloading the dataset </h1>

In [4]:
!wget -O seattle-data-collision.csv https://amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud/Full%20Data%20Collisions.csv

--2020-08-30 22:17:20--  https://amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud/Full%20Data%20Collisions.csv
Resolving amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud (amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud)... 169.63.118.98
Connecting to amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud (amapplied-data-science-capstone.s3.us-east.cloud-object-storage.appdomain.cloud)|169.63.118.98|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 84554087 (81M) [text/csv]
Saving to: ‘seattle-data-collision.csv’

100%[======================================>] 84,554,087  95.6MB/s   in 0.8s   

2020-08-30 22:17:21 (95.6 MB/s) - ‘seattle-data-collision.csv’ saved [84554087/84554087]



<h1> Understanding the Data </h1>

In [5]:
df = pd.read_csv('seattle-data-collision.csv')
df.shape

(221006, 40)

In [6]:
df.head()

X              Y  OBJECTID  INCKEY  COLDETKEY  REPORTNO  \
0  1.268969e+06  231759.927713         1  333240     334740   3851889   
1  1.271618e+06  202779.935236         2  333317     334817   3834541   
2  1.271322e+06  217404.090003         3    1367       1367   3671783   
3  1.269294e+06  224852.790938         4    1189       1189   3548948   
4  1.259501e+06  255054.109113         5   30400      30400  04046023   

      STATUS      ADDRTYPE   INTKEY  \
0  Unmatched  Intersection  28743.0   
1  Unmatched         Block      NaN   
2    Matched  Intersection  31348.0   
3    Matched         Block      NaN   
4  Unmatched  Intersection  23863.0   

                                       LOCATION  ... ROADCOND LIGHTCOND  \
0                          9TH AVE N AND ROY ST  ...      NaN       NaN   
1  S MICHIGAN ST BETWEEN 5TH PL S AND 6TH AVE S  ...      NaN       NaN   
2                    4TH AVE S AND S HOLGATE ST  ...      Dry  Daylight   
3   1ST AVE BETWEEN SENECA ST AND UNIVERSITY ST  ...      Dry  Daylight   
4                    18TH AVE NW AND NW 83RD ST  ...      NaN       NaN   

  PEDROWNOTGRNT SDOTCOLNUM SPEEDING  ST_COLCODE  \
0           NaN        NaN      NaN               
1           NaN        NaN      NaN               
2           NaN        NaN      NaN          13   
3           NaN        NaN      NaN          11   
4           NaN  4046023.0      NaN         NaN   

                                          ST_COLDESC  SEGLANEKEY  \
0                                                NaN           0   
1                                                NaN           0   
2  From same direction - both going straight - bo...           0   
3  From same direction - both going straight - bo...           0   
4                                                NaN           0   

   CROSSWALKKEY  HITPARKEDCAR  
0             0             N  
1             0             N  
2             0             N  
3             0             N  
4             0             N  

[5 rows x 40 columns]

<h2> Checking all columns for Missing Values </h2>

In [7]:
missing_data=df.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

X
False    213539
True       7467
Name: X, dtype: int64

Y
False    213539
True       7467
Name: Y, dtype: int64

OBJECTID
False    221006
Name: OBJECTID, dtype: int64

INCKEY
False    221006
Name: INCKEY, dtype: int64

COLDETKEY
False    221006
Name: COLDETKEY, dtype: int64

REPORTNO
False    221006
Name: REPORTNO, dtype: int64

STATUS
False    221006
Name: STATUS, dtype: int64

ADDRTYPE
False    217295
True       3711
Name: ADDRTYPE, dtype: int64

INTKEY
True     149278
False     71728
Name: INTKEY, dtype: int64

LOCATION
False    216422
True       4584
Name: LOCATION, dtype: int64

EXCEPTRSNCODE
True     120403
False    100603
Name: EXCEPTRSNCODE, dtype: int64

EXCEPTRSNDESC
True     209237
False     11769
Name: EXCEPTRSNDESC, dtype: int64

SEVERITYCODE
False    221005
True          1
Name: SEVERITYCODE, dtype: int64

SEVERITYDESC
False    221006
Name: SEVERITYDESC, dtype: int64

COLLISIONTYPE
False    194678
True      26328
Name: COLLISIONTYPE, dtype: int64

PERSONCOUNT
False    22

#### Cleaning Data for columns with no relevant usage

Drop columns X, Y, OBJECTID, INCKEY, COLDETKEY, REPORTNO, EXCEPTRSNCODE, EXCEPTRSNDESC, SEGLANEKEY, CROSSWALKKEY, INCDTTM, SDOTCOLNUM, ST_COLCODE, ST_COLDESC, HITPARKEDCAR, SDOT_COLDESC, SDOT_COLCODE, LOCATION, INTKEY as these are primary key columns and are not providing any new information for the algorithm


In [8]:
df.drop(['X','Y','OBJECTID','INCKEY','COLDETKEY','REPORTNO','EXCEPTRSNCODE','EXCEPTRSNDESC','SEGLANEKEY','CROSSWALKKEY',
         'INCDTTM','SDOTCOLNUM','ST_COLCODE','ST_COLDESC','HITPARKEDCAR','SDOT_COLDESC','SDOT_COLCODE','LOCATION','INTKEY'], axis=1, inplace=True)

In [9]:
df.head()

STATUS      ADDRTYPE SEVERITYCODE                    SEVERITYDESC  \
0  Unmatched  Intersection            2                Injury Collision   
1  Unmatched         Block            2                Injury Collision   
2    Matched  Intersection            1  Property Damage Only Collision   
3    Matched         Block            1  Property Damage Only Collision   
4  Unmatched  Intersection            0                         Unknown   

  COLLISIONTYPE  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  VEHCOUNT  INJURIES  ...  \
0           NaN            2         0            0         0         2  ...   
1           NaN            3         0            0         0         1  ...   
2    Rear Ended            2         0            0         2         0  ...   
3     Sideswipe            8         0            0         2         0  ...   
4           NaN            0         0            0         0         0  ...   

   FATALITIES                 INCDATE  \
0           0  2020/08/10 00:00:00+00   
1           0  2020/08/14 00:00:00+00   
2           0  2013/04/02 00:00:00+00   
3           0  2013/03/30 00:00:00+00   
4           0  2004/02/15 00:00:00+00   

                              JUNCTIONTYPE INATTENTIONIND UNDERINFL WEATHER  \
0   At Intersection (intersection related)            NaN       NaN     NaN   
1  Mid-Block (not related to intersection)            NaN       NaN     NaN   
2   At Intersection (intersection related)            NaN         N   Clear   
3  Mid-Block (not related to intersection)            NaN         N   Clear   
4   At Intersection (intersection related)            NaN       NaN     NaN   

  ROADCOND LIGHTCOND PEDROWNOTGRNT SPEEDING  
0      NaN       NaN           NaN      NaN  
1      NaN       NaN           NaN      NaN  
2      Dry  Daylight           NaN      NaN  
3      Dry  Daylight           NaN      NaN  
4      NaN       NaN           NaN      NaN  

[5 rows x 21 columns]

#### Check for missing values in the updated dataset 

In [10]:
df.head()
missing_data=df.isnull()
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

STATUS
False    221006
Name: STATUS, dtype: int64

ADDRTYPE
False    217295
True       3711
Name: ADDRTYPE, dtype: int64

SEVERITYCODE
False    221005
True          1
Name: SEVERITYCODE, dtype: int64

SEVERITYDESC
False    221006
Name: SEVERITYDESC, dtype: int64

COLLISIONTYPE
False    194678
True      26328
Name: COLLISIONTYPE, dtype: int64

PERSONCOUNT
False    221006
Name: PERSONCOUNT, dtype: int64

PEDCOUNT
False    221006
Name: PEDCOUNT, dtype: int64

PEDCYLCOUNT
False    221006
Name: PEDCYLCOUNT, dtype: int64

VEHCOUNT
False    221006
Name: VEHCOUNT, dtype: int64

INJURIES
False    221006
Name: INJURIES, dtype: int64

SERIOUSINJURIES
False    221006
Name: SERIOUSINJURIES, dtype: int64

FATALITIES
False    221006
Name: FATALITIES, dtype: int64

INCDATE
False    221006
Name: INCDATE, dtype: int64

JUNCTIONTYPE
False    209046
True      11960
Name: JUNCTIONTYPE, dtype: int64

INATTENTIONIND
True     190818
False     30188
Name: INATTENTIONIND, dtype: int64

UNDERINFL
False    194698

### Correcting Data or removing missing rows

In order to correct the dataset we need to look at the dataset for each column and convert the values appropriately

Following Columns are not missing any values and need no correction
1. STATUS
2. SEVERITYDESC
3. PERSONCOUNT
4. PEDCOUNT
5. PEDCYLCOUNT
6. VEHCOUNT
7. INJURIES
8. SERIOUSINJURIES
9. FATALITIES
10. INCDATE

Let's look at ADDRTYPE

In [ ]:
df['ADDRTYPE'].value_counts()

#### Block - 144694, Alley - 873, Intersection - 71728


You can now convert rest of the records to addresstype as Unknown

In [15]:
df['ADDRTYPE'].replace(np.nan, 'Unknown', inplace=True)

Let's now look at another column - SEVERITYCODE <br>
According to the metadata:
A code that corresponds to the severity of the collision: <br>
• 3—fatality <br>
• 2b—serious injury<br>
• 2—injury<br>
• 1—prop damage<br>
• 0—unknown<br>

In [21]:
df.loc[df['SEVERITYCODE'].isnull()]

STATUS      ADDRTYPE SEVERITYCODE SEVERITYDESC COLLISIONTYPE  \
195468  Matched  Intersection          NaN      Unknown           NaN   

        PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  VEHCOUNT  INJURIES  ...  \
195468            2         0            1         1         0  ...   

        FATALITIES                 INCDATE  \
195468           0  2018/02/14 00:00:00+00   

                                  JUNCTIONTYPE INATTENTIONIND UNDERINFL  \
195468  At Intersection (intersection related)            NaN         N   

         WEATHER ROADCOND LIGHTCOND PEDROWNOTGRNT SPEEDING  
195468  Overcast      Wet  Daylight           NaN      NaN  

[1 rows x 21 columns]

Convert the missing Severity Code based on the description

In [22]:
df['SEVERITYCODE'].replace(np.nan, '0', inplace=True)

Let's look at one more column - INATTENTIONIND <br>
According to the metadata, this attritbue is describer as Whether or not collision was due to inattention. (Y/N)

In [30]:
df['INATTENTIONIND'].isnull().value_counts()

True     190818
False     30188
Name: INATTENTIONIND, dtype: int64

In [31]:
df['INATTENTIONIND'].value_counts()

Y    30188
Name: INATTENTIONIND, dtype: int64

Only values with Y are populated, rest all records are null, so we assume that this must be N

In [32]:
df['INATTENTIONIND'].replace(np.nan, 'N', inplace=True)

#### Similarly, we'll standardize all the missing records

In [33]:
df['SEVERITYCODE'].replace(np.nan, '0', inplace=True)
df['JUNCTIONTYPE'].replace(np.nan, 'Unknown', inplace=True)
df['ADDRTYPE'].replace(np.nan, 'Unknown', inplace=True)
df['COLLISIONTYPE'].replace(np.nan, 'Unknown', inplace=True)
df['INATTENTIONIND'].replace(np.nan, 'N', inplace=True)
df['UNDERINFL'].replace(np.nan, 'N', inplace=True)
df['UNDERINFL'].replace('0', 'N', inplace=True)
df['UNDERINFL'].replace('1', 'Y', inplace=True)
df['WEATHER'].replace(np.nan, 'Unknown', inplace=True)
df['ROADCOND'].replace(np.nan, 'Unknown', inplace=True)
df['LIGHTCOND'].replace(np.nan, 'Unknown', inplace=True)
df['PEDROWNOTGRNT'].replace(np.nan, 'N', inplace=True)
df['SPEEDING'].replace(np.nan,'N',inplace=True)
df['INCDATE'] = pd.to_datetime(df['INCDATE'],format='%Y-%m-%d')

In [34]:
for column in df.columns.values.tolist():
    print(column)
    print (df[column].value_counts())
    print("")    

STATUS
Matched      194751
Unmatched     26255
Name: STATUS, dtype: int64

ADDRTYPE
Block           144694
Intersection     71728
Unknown           3711
Alley              873
Name: ADDRTYPE, dtype: int64

SEVERITYCODE
1     137335
2      58628
0      21602
2b      3092
3        349
Name: SEVERITYCODE, dtype: int64

SEVERITYDESC
Property Damage Only Collision    137335
Injury Collision                   58628
Unknown                            21602
Serious Injury Collision            3092
Fatality Collision                   349
Name: SEVERITYDESC, dtype: int64

COLLISIONTYPE
Parked Car    48446
Angles        35444
Rear Ended    34609
Unknown       26328
Other         24512
Sideswipe     18849
Left Turn     14075
Pedestrian     7652
Cycles         5905
Right Turn     3006
Head On        2180
Name: COLLISIONTYPE, dtype: int64

PERSONCOUNT
2     118788
3      36451
0      24995
4      14990
1      13882
5       6776
6       2785
7       1180
8        547
9        227
10       133
11    

In [35]:
df.head()

STATUS      ADDRTYPE SEVERITYCODE                    SEVERITYDESC  \
0  Unmatched  Intersection            2                Injury Collision   
1  Unmatched         Block            2                Injury Collision   
2    Matched  Intersection            1  Property Damage Only Collision   
3    Matched         Block            1  Property Damage Only Collision   
4  Unmatched  Intersection            0                         Unknown   

  COLLISIONTYPE  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  VEHCOUNT  INJURIES  ...  \
0       Unknown            2         0            0         0         2  ...   
1       Unknown            3         0            0         0         1  ...   
2    Rear Ended            2         0            0         2         0  ...   
3     Sideswipe            8         0            0         2         0  ...   
4       Unknown            0         0            0         0         0  ...   

   FATALITIES                   INCDATE  \
0           0 2020-08-10 00:00:00+00:00   
1           0 2020-08-14 00:00:00+00:00   
2           0 2013-04-02 00:00:00+00:00   
3           0 2013-03-30 00:00:00+00:00   
4           0 2004-02-15 00:00:00+00:00   

                              JUNCTIONTYPE INATTENTIONIND UNDERINFL  WEATHER  \
0   At Intersection (intersection related)              N         N  Unknown   
1  Mid-Block (not related to intersection)              N         N  Unknown   
2   At Intersection (intersection related)              N         N    Clear   
3  Mid-Block (not related to intersection)              N         N    Clear   
4   At Intersection (intersection related)              N         N  Unknown   

  ROADCOND LIGHTCOND PEDROWNOTGRNT SPEEDING  
0  Unknown   Unknown             N        N  
1  Unknown   Unknown             N        N  
2      Dry  Daylight             N        N  
3      Dry  Daylight             N        N  
4  Unknown   Unknown             N        N  

[5 rows x 21 columns]

In [38]:
df.groupby(['ADDRTYPE'])['SEVERITYDESC'].value_counts(normalize=True)

ADDRTYPE      SEVERITYDESC                  
Alley         Property Damage Only Collision    0.772050
              Unknown                           0.128293
              Injury Collision                  0.093929
              Serious Injury Collision          0.005727
Block         Property Damage Only Collision    0.673228
              Injury Collision                  0.209663
              Unknown                           0.105264
              Serious Injury Collision          0.010519
              Fatality Collision                0.001327
Intersection  Property Damage Only Collision    0.522976
              Injury Collision                  0.390615
              Unknown                           0.062444
              Serious Injury Collision          0.021777
              Fatality Collision                0.002189
Unknown       Unknown                           0.479655
              Property Damage Only Collision    0.468068
              Injury Collision             

### 77% collisions/accidents in alley results in property damage while ~40% collisions/accidents happening at Intersection results in Injury

In [37]:
df.groupby(['UNDERINFL'])['SEVERITYDESC'].value_counts(normalize=True)

UNDERINFL  SEVERITYDESC                  
N          Property Damage Only Collision    0.623403
           Injury Collision                  0.260511
           Unknown                           0.102197
           Serious Injury Collision          0.012679
           Fatality Collision                0.001211
Y          Property Damage Only Collision    0.577630
           Injury Collision                  0.369924
           Serious Injury Collision          0.042787
           Fatality Collision                0.009658
Name: SEVERITYDESC, dtype: float64

### 10% more chances of injury in a collision when the driver was under influence of alcohol